File: Binary_Search_Tree/javascript/delete_node.js
// JavaScript implementation
// Delete Operation


// Visual Representation of the binary search tree
//                10
//           /          \
//         5              17
//      /    \           /   \
//     2      7        14     22
//    / \    / \      / \     / \
//   1   4  6   9   11   16 22   27

// Node class
class Node
{
    constructor(data)
    {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}



// Binary Search tree class
class BinarySearchTree
{
    constructor()
    {
        this.root = null;
    }

    // Insert Operation
    insert(data)
    {
        var newNode = new Node(data);
        if(this.root === null)
            this.root = newNode;
        else
            this.insertNode(this.root, newNode);
    }

    insertNode(node, newNode)
    {
        if(newNode.data < node.data)
        {
            if(node.left === null)
                node.left = newNode;
            else
                this.insertNode(node.left, newNode); 
        }

        else
        {
            if(node.right === null)
                node.right = newNode;
            else
                this.insertNode(node.right,newNode);
        }
    }

    // Delete Operation
    remove(data)
    {
        this.root = this.removeNode(this.root, data);
    }


    removeNode(node, key)
    {
        if(node === null)
            return null;
        else if(key < node.data)
        {
            node.left = this.removeNode(node.left, key);
            return node;
        }
        else if(key > node.data)
        {
            node.right = this.removeNode(node.right, key);
            return node;
        }
        else
        {
            if(node.left === null && node.right === null)
            {
                node = null;
                return node;
            }

            if(node.left === null)
            {
                node = node.right;
                return node;
            }
            else if(node.right === null)
            {
                node = node.left;
                return node;
            }
            var aux = this.findMinNode(node.right);
            node.data = aux.data;
            node.right = this.removeNode(node.right, aux.data);
            return node;
        }
    }



    // To print in inoder method
    inorder(node)
    {
        if(node !== null)
        {
            this.inorder(node.left);
            console.log(node.data);
            this.inorder(node.right);
        }
    }


    findMinNode(node)
    {
        if(node.left === null)
            return node;
        else
            return this.findMinNode(node.left);
    }


    getRootNode()
    {
        return this.root;
    }

    // Searching Element
    search(node, data)
    {
        if(node === null)
            return null;

        else if(data < node.data)
            return this.search(node.left, data);

        else if(data > node.data)
            return this.search(node.right, data);

        else
            return node;
    }

}







var BST = new BinarySearchTree();

// Inserting Elements
BST.insert(10);

BST.insert(2);
BST.insert(1);

BST.insert(4);
BST.insert(7);

BST.insert(9);
BST.insert(6);

BST.insert(5);
BST.insert(11);

BST.insert(14);
BST.insert(19);

BST.insert(17);
BST.insert(16);

BST.insert(27);
BST.insert(22);



var root = BST.getRootNode();
console.log('\n');
BST.inorder(root);
console.log('\n');

// var a = BST.search(root,5)


// if(a === null)
//     console.log('\nNot Found');
// else if(a.data !== 5)
//     console.log('\nNot Found');
// else 
//     console.log('\nFound');

BST.remove(9);
console.log('\n');
BST.inorder(root);

BST.remove(7);
console.log('\n');
BST.inorder(root);

BST.remove(14);
console.log('\n');
BST.inorder(root);

File: ACADEMIC CODES/Binary_Search_Tree/java/delete_node.java
// Binary Search Tree 
// JAVA

//  Visual Representation of the binary search tree
//                 10
//            /         \ 
//          5              17
//       /    \           /   \ 
//      2      7        14     22
//     / \    / \      / \     / \       
//    1   4  6   9   11   16 22   27  

class BinarySearchTree {
    class Node {
      int key;
      Node left, right;
  
      public Node(int item) {
        key = item;
        left = right = null;
      }
    }
  
    Node root;
  
    BinarySearchTree() {
      root = null;
    }
  
    void insert(int key) {
      root = insertKey(root, key);
    }
  
    Node insertKey(Node root, int key) {
      if (root == null) {
        root = new Node(key);
        return root;
      }
      if (key < root.key)
        root.left = insertKey(root.left, key);
      else if (key > root.key)
        root.right = insertKey(root.right, key);
  
      return root;
    }
  
    void inorder() {
      inorderRec(root);
    }
  
    void inorderRec(Node root) {
      if (root != null) {
        inorderRec(root.left);
        System.out.print(" -> " + root.key);
        inorderRec(root.right);
      }
    }
  
    void deleteKey(int key) {
      root = deleteRec(root, key);
    }
  
    Node deleteRec(Node root, int key) {
      if (root == null)
        return root;
      if (key < root.key)
        root.left = deleteRec(root.left, key);
      else if (key > root.key)
        root.right = deleteRec(root.right, key);
      else {
        if (root.left == null)
          return root.right;
        else if (root.right == null)
          return root.left;
        root.key = minValue(root.right);
        root.right = deleteRec(root.right, root.key);
      }
  
      return root;
    }
  

    int minValue(Node root) {
      int minv = root.key;
      while (root.left != null) {
        minv = root.left.key;
        root = root.left;
      }
      return minv;
    }
  
    public static void main(String[] args) {
      BinarySearchTree tree = new BinarySearchTree();
  
      tree.insert(10);

      tree.insert(2);
      tree.insert(1);
      tree.insert(4);
      tree.insert(9);

      tree.insert(7);
      tree.insert(6);
      tree.insert(5);
      tree.insert(9);
      tree.insert(11);
      tree.insert(14);
      tree.insert(19);
      tree.insert(17);

      tree.insert(16);
      tree.insert(27);
      tree.insert(22);

  
      System.out.print("Inorder traversal: ");
      tree.inorder();
  
      System.out.println("\n\nAfter deleting 10");
      tree.deleteKey(10);
      System.out.print("Inorder traversal: ");
      tree.inorder();
    }
  }
